includes/Abilities/ExecuteAbilityAbility.php

```diff
@@ -0,0 +1,222 @@
+		$user_check = self::validate_user_access();
+		if ( is_wp_error( $user_check ) ) {
+			return $user_check;
+		}
+
+		// Check MCP exposure restrictions
+		$exposure_check = self::check_ability_mcp_exposure( $ability_name );
+		if ( is_wp_error( $exposure_check ) ) {
+			return $exposure_check;
+		}
+
+		// Get the target ability
+		$ability = wp_get_ability( $ability_name );
+		if ( ! $ability ) {
+			return new \WP_Error( 'ability_not_found', "Ability '{$ability_name}' not found" );
+		}
+
+		// Check if the user has permission to execute the target ability
+		$parameters        = $input['parameters'] ?? array();
+		$permission_result = $ability->check_permissions( $parameters );
+
+		// Return WP_Error as-is, or convert other values to boolean
+		if ( is_wp_error( $permission_result ) ) {
+			return $permission_result;
+		}
+
+		return (bool) $permission_result;
+	}
+
+	/**
+	 * Validate user authentication and basic capabilities for execute ability.
+	 *
+	 * @return bool|\WP_Error True if valid, WP_Error if validation fails.
+	 */
+	private static function validate_user_access() {
+		// Verify caller identity - ensure user is authenticated
+		if ( ! is_user_logged_in() ) {
+			return new \WP_Error( 'authentication_required', 'User must be authenticated to access this ability' );
+		}
+
+		// Check basic capability requirement - allow customization via filter
+		$required_capability = apply_filters( 'mcp_adapter_execute_ability_capability', 'read' );
+		// phpcs:ignore WordPress.WP.Capabilities.Undetermined -- Capability is determined dynamically via filter
+		if ( ! current_user_can( $required_capability ) ) {
+			return new \WP_Error(
+				'insufficient_capability',
+				sprintf( 'User lacks required capability: %s', $required_capability )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Execute the ability execution functionality.
+	 *
+	 * Enforces security checks before executing any ability.
+	 *
+	 * @param array $input Input parameters containing ability_name and parameters.
+	 *
+	 * @return array Array containing execution results.
+	 */
+	public static function execute( $input = array() ): array {
+		$ability_name = $input['ability_name'] ?? '';
+		$parameters   = $input['parameters'] ?? array();
+
+		if ( empty( $ability_name ) ) {
+			return array(
+				'success' => false,
+				'error'   => 'Ability name is required',
+			);
+		}
+
+		// Enforce security checks before execution
+		// Note: WordPress will have already called check_permission, but we double-check
+		// as an additional security layer for direct method calls
+		$permission_check = self::check_permission( $input );
+		if ( is_wp_error( $permission_check ) ) {
+			return array(
+				'success' => false,
+				'error'   => $permission_check->get_error_message(),
+			);
+		}
+
+		if ( ! $permission_check ) {
+			return array(
+				'success' => false,
+				'error'   => 'Permission denied for ability execution',
+			);
+		}
```

Comment:
Hmm, I don't love the idea that abilities have to check their own permissions in special circumstances. If someone programmatically bypasses the permission check to run an ability, then that should just work. They can obviously otherwise bypass it via a filter.